{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1k6EJO8Zjn4rh6qdKZQl549J-W0CRLLS8","timestamp":1741376200516}]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"TiM6gYg0nhkY"},"source":["<font color=\"#de3023\"><h1><b>REMINDER MAKE A COPY OF THIS NOTEBOOK, DO NOT EDIT</b></h1></font>"]},{"cell_type":"markdown","metadata":{"id":"XFYAV890_btm"},"source":["# Goals\n","In this colab you will:\n","* Brainstorm how to use AI for sustainability and conservation, and to combat climate change.\n","* Learn the ways in which naturally occurring bacteria can help increase plant growth and reforestation.\n","* Examine a dataset, and build a machine learning model to predict crop yield from soil composition.\n","* OPTIONALLY: Learn how scientists collect data about bacterial composition of an environment & measure the number of different bacteria in soils using DNA sequences."]},{"cell_type":"code","metadata":{"id":"VoQM7javZl7b"},"source":["#@title ### Setup notebook.\n","\n","# Sample metadata\n","!wget -q --show-progress \"https://storage.googleapis.com/inspirit-ai-data-bucket-1/Data/AI%20Scholars/Sessions%206%20-%2010%20(Projects)/Project%20-%20Sustainable%20Farming/sample_metadata.tsv\"\n","\n","# 16S_counts\n","!wget -q --show-progress \"https://storage.googleapis.com/inspirit-ai-data-bucket-1/Data/AI%20Scholars/Sessions%206%20-%2010%20(Projects)/Project%20-%20Sustainable%20Farming/16S_counts.tsv\"\n","\n","# bacteria_counts\n","!wget -q --show-progress \"https://storage.googleapis.com/inspirit-ai-data-bucket-1/Data/AI%20Scholars/Sessions%206%20-%2010%20(Projects)/Project%20-%20Sustainable%20Farming/bacteria_counts.tsv\"\n","\n","# sequence_to_species_dict\n","!wget -q --show-progress \"https://storage.googleapis.com/inspirit-ai-data-bucket-1/Data/AI%20Scholars/Sessions%206%20-%2010%20(Projects)/Project%20-%20Sustainable%20Farming/sequence_to_species_dict.npy\"\n","\n","import pandas as pd\n","import numpy as np\n","import seaborn as sns\n","import matplotlib.pyplot as plt\n","from sklearn.model_selection import train_test_split\n","from sklearn.tree import DecisionTreeRegressor\n","from sklearn.tree import plot_tree\n","\n","metadata = pd.read_table('sample_metadata.tsv')\n","metadata.index = ['farm_%i' %i for i in range(len(metadata))]\n","\n","sequences_counts = pd.read_table('16S_counts.tsv')\n","sequences_counts.index = ['farm_%i' % i for i in range(len(sequences_counts))]\n","\n","bacteria_counts = pd.read_table('bacteria_counts.tsv')\n","bacteria_counts.index = ['farm_%i' % i for i in range(len(bacteria_counts))]\n","cols = list(bacteria_counts.columns)\n","np.random.seed(42)\n","np.random.shuffle(cols)\n","bacteria_counts = bacteria_counts[cols]\n","sequence_to_species_dict = np.load('sequence_to_species_dict.npy', allow_pickle=True).item()\n","\n","bacteria_counts = bacteria_counts.drop(['Unnamed: 0'], axis=1)\n","print(\"Setup successful.\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"t09XQ0aZcUFT"},"source":["# People Health and Planet Health\n","\n","<img src=\"https://www.gstatic.com/earth/social/00_generic_facebook-001.jpg\" alt=\"drawing\" width=\"1000\" height=\"300\"/>\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"G4dOO56rISWQ"},"source":["In almost all of the problems we've worked on in this course so far, the goal is to improve the lives and health of people around the world (and dogs that have tendencies to run into the street!).  But we're forgetting the health of someone else pretty important.... Mother Earth!\n","\n"," **Discuss: What does planet health mean to you?**\n","\n","Some people feel that people health and planet health are sometimes at odds with each other.  That is, people should make sacrifices (like driving less, opt to buy more expensive but sustainable goods, etc.) in order to save the planet. Others think that people health and planet health go hand-in-hand. A healthier global ecosystem means a healthier human!  \n","\n","####**Discuss: Which do you think?**\n","\n","\n","#### **Exercise: Brainstorm one problem that pits planet health *against* people health, and one problem where planet health and people health go hand-in-hand.**\n"]},{"cell_type":"code","metadata":{"id":"x7L85UXAIR4i","cellView":"form"},"source":["people_vs_planet_ = '' #@param {type:\"string\"}\n","hand_in_hand_ = '' #@param {type:\"string\"}\n","print('Example answers: \\n Plane travel is extremely carbon-expensive, \\n but many people enjoy travelling or visiting people far away.  \\n On the other hand, water pollution affects the health of local \\n animals, plants, and humans.')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"89LAMV1yLkqv"},"source":["# Sustainable Farming and AI\n","\n","One way we can improve both people and planet health is through better farming practices. More efficient and effective farming practices can have greater crop yields while using less land, energy, and harmful chemicals.  The world is projected to have 9 billion people by 2050... How can we make sure we are able to nourish 9 billion people while also nourishing our planet?\n","\n","If you haven't already done so, please read [this](https://www.nationalgeographic.com/foodfeatures/feeding-9-billion/) article on 5 proposed steps to sustainably feeding the world.\n"]},{"cell_type":"markdown","metadata":{"id":"rPZMbDbMM67b"},"source":["\"*Just as electricity transformed almost everything 100 years ago, today I actually have a hard time thinking of an industry that I don't think AI will transform in the next several years.*\" -- Andrew Ng\n","\n","\n","So, why not planet earth and sustainable farming too?\n","\n","#### **Exercise: Brainstorm a way AI could aid in each of the 5 proposed steps to sustainable farming.**\n"]},{"cell_type":"code","metadata":{"id":"hN2HwoLbMvaq","cellView":"form"},"source":["_1_ = '' #@param {type:\"string\"}\n","_2_ = '' #@param {type:\"string\"}\n","_3_ = '' #@param {type:\"string\"}\n","_4_ = '' #@param {type:\"string\"}\n","_5_ = '' #@param {type:\"string\"}\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"B-zehrULZdnb"},"source":["# Let's look at some farms!\n","\n","<img src=\"https://cdn.pixabay.com/photo/2017/08/02/10/31/farming-2570803_1280.jpg\" width=500>\n","\n","Here is a dataset with crop yields and other information from different farms (stored in the variable ```metadata```).  Go ahead and explore the dataset. We've written a little bit of code to get you started.\n"]},{"cell_type":"code","metadata":{"id":"_y882J5cZx-G"},"source":["metadata.head()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"xF8His7pXY93"},"source":["####**Exercise: Using the metadata & pandas, answer the following:**\n","1. Where did these samples come from?\n","2. For what type of plant was the yield measured?\n","3. Were all of these samples collected on the same day?"]},{"cell_type":"code","metadata":{"id":"tJmO_DQ5O0TV","cellView":"form"},"source":["_1_ = '' #@param {type:\"string\"}\n","_2_ = '' #@param {type:\"string\"}\n","_3_ = '' #@param {type:\"string\"}\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Instructor Solution\n","<details><summary>click to reveal!</summary>\n","\n","Answers: Australia, Barley, No"],"metadata":{"id":"kL2cmLeyaDZ7"}},{"cell_type":"markdown","metadata":{"id":"9I6AsU-hObEI"},"source":["### **Exercise: Plot a histogram of the crop yields.  What do you observe from the histogram?**\n","\n","Take a look at [the `histplot` documentation](https://seaborn.pydata.org/generated/seaborn.histplot.html) and play around with some of the optional parameters. How does your graph change?"]},{"cell_type":"code","metadata":{"id":"hnHf_iHNg3n2"},"source":["ax = sns.histplot(data=metadata['crop_yield'])\n","ax.set(xlabel=None, ylabel=None) ### FILL IN ###\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ABopRLnsQMpM"},"source":["#@title ###Example Instructor Solution\n","# plt.hist(metadata['crop_yield'])\n","# plt.xlabel('Crop Yield')\n","# plt.ylabel('Number of Farms')\n","# plt.show()\n","\n","ax = sns.histplot(data=metadata['crop_yield'])\n","ax.set(xlabel='Crop Yield (Kilograms/hectare)', ylabel='Number of Farms')\n","plt.show()\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"glaeSfvuOlCY"},"source":["In this class we are going to use some interesting measurements (I won't spoil what these are just yet!) in order to predict how well plants or crops will grow (called *yield*) in a certain locale.\n","\n","**Discuss: What \"step\" in the previous article would this fall under?**\n","\n","####**Exercise: Brainstorm 3 attributes of soil (ie *features*) we could measure to predict crop yield?** It's ok if you aren't a chemistry or plant biology whiz -- get creative! Anything goes!\n"]},{"cell_type":"code","metadata":{"id":"nnsDEhtNad1Q","cellView":"form"},"source":["_1_ = '' #@param {type:\"string\"}\n","_2_ = '' #@param {type:\"string\"}\n","_3_ = '' #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ThAOo5-JQt42"},"source":["# Predicting Crop Yield From....\n","\n","\n","\n","<img src=\"https://t4.ftcdn.net/jpg/01/30/73/43/360_F_130734330_mSBGYTwuGvaHhEoyHlDov6ELFHO4cexF.jpg\" alt=\"drawing\" width=\"1000\" height=\"500\"/>\n","\n","\n","\n","#.... Bacteria!\n","\n"]},{"cell_type":"markdown","metadata":{"id":"bsdbCJ9XRgkh"},"source":["\n","\n","Scientists have discovered that the soil microbiome, the collection of bacteria that live in a region of soil play an important role in the health of plants! (Interestingly, the bacteria that live in the human intestines also play an important role in the health of humans!). Therefore, maybe we can predict how well plants will grow in a region based on the bacterial composition of the soil.\n","\n","Here is a short [tutorial](https://www.khanacademy.org/science/biology/bacteria-archaea/prokaryote-metabolism-ecology/a/prokaryote-interactions-ecology) on the role of prokaryotes (bacteria) in ecosystems that you can read if you want to learn more.\n","\n","\n","So, maybe we actually need to think *smaller* when it comes to sustainabity!  Like, so small you need a microscope ;)\n","\n","<img src=\"https://st2.depositphotos.com/1967477/6350/v/950/depositphotos_63507731-stock-illustration-cartoon-bacteria-collection-set.jpg\n","\" alt=\"drawing\" width=\"300\" >\n"]},{"cell_type":"markdown","metadata":{"id":"wBgrLUtccBu6"},"source":["# Examining Our Dataset\n","\n","Let's take a look at what these bacteria composition datasets look like! Go ahead and run the following line of code to examine your dataset."]},{"cell_type":"code","metadata":{"id":"M5-lA_j-cBME"},"source":["bacteria_counts.head()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"v_dOsmfCdJ6L"},"source":["#### **Exercise: What do the rows and columns represent?**"]},{"cell_type":"code","metadata":{"id":"lzwcV6r_dbZw","cellView":"form"},"source":["rows = '' #@param {type:\"string\"}\n","columns = '' #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KjraTFaPdiMO"},"source":["#### **Exercise: Do a quick google search on 3 bacteria from your dataset. Write a brief description of the function & environment of each bacteria.**\n","\n","Type your findings in the cell below, but no need to run the cell!"]},{"cell_type":"code","metadata":{"id":"k6p4ggBUdo_W","cellView":"form"},"source":["_1_ = '' #@param {type:\"string\"}\n","_2_ = '' #@param {type:\"string\"}\n","_3_ = '' #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"cLzGv1S2S_zL"},"source":["# Cleaning Up The Data\n","\n","\n","<img src=\"https://live.staticflickr.com/7040/6878034144_02a2e37731_b.jpg\" width=300>\n","\n","\n","Before building our machine learning model, we need to do a bit of preprocessing and data cleaning.\n","\n","We will do 2 things in order to clean up our data:\n","\n","1) We will get rid of any bacteria that are very low prevalence.\n","\n","2) We will log-normalize our data."]},{"cell_type":"markdown","metadata":{"id":"kBosBbx6uGtV"},"source":["## Removing Low Prevalence Bacteria\n","\n","If a bacteria is so rare that it appears in $<$10 samples, that bacteria is not going to be helpful for building machine learning models that generalize to new data. We call these **singletons**. In order to reduce the amount of data we make the model crunch through, we will remove any features/bacteria that are low-prevalence.\n"]},{"cell_type":"markdown","metadata":{"id":"J7PSd6bbVSh8"},"source":["###**Exercise: Fill out the code below to find the names of the bacteria that only appear in a single sample, and then remove them from your data.**"]},{"cell_type":"code","metadata":{"id":"ifZr-D1rVR2F"},"source":["low_prev_bacteria = []\n","bacteria = bacteria_counts.columns\n","for b in bacteria:\n","  if sum(bacteria_counts[b]>0)<10:  # If bacteria 'b' is a singleton,\n","\n","    # Fill in the code below to add bacteria 'b' to the list of singletons.\n","    low_prev_bacteria.append(____) ### FILL IN ###\n","\n","\n","# Fill in the code below to count the total number of singletons.\n","n_low_prevalence = 0000 #### FILL IN ###\n","print('%i bacteria are low prevalence.' % n_low_prevalence)\n","\n","# The 'drop' function drops the columns specified from the bacteria_counts dataframe.\n","bacteria_counts_no_low_prev = bacteria_counts.drop(low_prev_bacteria, axis=1)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"IYOtIS0inMKz"},"source":["#@title #### Example Solution.\n","low_prev_bacteria = []\n","bacteria = bacteria_counts.columns\n","for b in bacteria:\n","  if sum(bacteria_counts[b]>0)<10:  # If bacteria 'b' is a singleton,\n","\n","    # Fill in the code below to add bacteria 'b' to the list of singletons.\n","    low_prev_bacteria.append(b)\n","\n","\n","# Fill in the code below to count the total number of singletons.\n","n_low_prevalence = len(low_prev_bacteria)\n","print('%i bacteria are low prevalence.' % n_low_prevalence)\n","\n","# The 'drop' function drops the columns specified from the bacteria_counts dataframe.\n","bacteria_counts_no_low_prev = bacteria_counts.drop(low_prev_bacteria, axis=1)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KPTw4YPluKfa"},"source":["## Log-normalizing\n","\n","**Discuss the following question: An apple tree grows 100 apples one year, and then 110 apples the next year.  Meanwhile, a pumpkin vine grows 10 pumpkins one year, and then 20 pumpkins the next year. Which plant do you feel like experienced a greater change between years?**\n","\n","Although the plants had the same change in *magnitude* of fruits, you probably feel like the pumpkin plant experienced a greater change because it doubled in fruit production, while the apple tree only increased fruit production by 10%. The pumpkin vine has a larger *relative* change year-to-year. In biology, we are often interested in relative changes as well. A trick for looking at relative changes in data is to use log normalization.\n","\n","To log-normalize a number N, we do:\n","\n","$N_{norm} = log(N+1)$\n","\n","To take the log of a number in our code, we can use the `np.log` function. Take a look at [the documentation for the function](https://numpy.org/doc/stable/reference/generated/numpy.log.html) if you'd like to see some examples of how it's used!\n","\n"]},{"cell_type":"markdown","metadata":{"id":"NLakcUiufbhn"},"source":["**Optional: Why do you think we also add 1? (Hint: What would happen if N equaled 0?)**\n","\n","###**Exercise: Perform log-normalization on your dataset by filling out the following code.**\n"]},{"cell_type":"code","metadata":{"id":"5s1hygHTuKGM"},"source":["# Fill out the following one-liner to log-normalize your data. Hint: numpy is\n","# your friend! Remember, you should be performing log-normalization on the\n","# bacteria_counts_no_low_prev dataframe.\n","bacteria_counts_lognorm = None #### FILL THIS OUT."],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"gP22IId7oP5W"},"source":["#@title #### Example Solution\n","# Fill out the following one-liner to log-normalize your data. Hint: numpy is\n","# your friend! Remember, you should be performing log-normalization on the\n","# bacteria_counts_no_low_prev dataframe.\n","bacteria_counts_lognorm = np.log(bacteria_counts_no_low_prev + 1)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ioNbvvN0na9Z"},"source":["bacteria_counts_lognorm.to_csv('bacteria_counts_lognorm.csv')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"rX8cShiMkWQ7"},"source":["# Wrapping Up\n","\n","Great job! Data cleaning and exploration is an important part of machine learning especially when it comes to working with biological or environmental data!\n","\n","In the next notebook, we will be building a model in order to predict crop yield from bacterial composition.\n","\n","####**Exericse: To review what you learned in the class answer the following questions:**\n","1. Is this a regression or classification problem?\n","2. What will be the features (X data / input variable) in our model?\n","3. What will be the label (Y data / output variable) in our model?\n","\n","\n"]},{"cell_type":"code","metadata":{"id":"30UQggPrlmBU","cellView":"form"},"source":["_1_ = '' #@param {type:\"string\"}\n","_2_ = '' #@param {type:\"string\"}\n","_3_ = '' #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JopNFQXGmR8w"},"source":["#### **And finally, discuss: how can this model help sustainable farming in the face of climate change?**\n","\n"]},{"cell_type":"markdown","metadata":{"id":"0JqESxGdTECl"},"source":["# (Optional) Data Exploration\n","\n","We are going to use a method of unsupervised learning called **hierarchical clustering** in order to explore and visualize our data. Unsupervised learning is usually used by machine learning scientists to understand what our features look like (if any features are correlated, or if there are certain outliers in the dataset) before diving into machine learning models.\n","\n","\n","Hierarchical clustering groups together features that occur in similar samples, and also clusters together samples whose features look similar.\n","\n","Unsupervised learning (and any unsupervised learning methods) do not use labels.  We only give the computer a set of features, and it will group together samples and features that look similar.\n","\n","\n","Here is an example of a hierarchical clustered dataset. ```is_vegan```, ```is_vegetarian```, etc. are features. Bacon, eggs, toast, etc. are the samples. Red means a value of 1, blue means a value of 0.\n","\n","\n","\n"," <img src=\"https://i.ibb.co/Fx0wMkm/hierical-clustering.png\n","\" alt=\"drawing\" width=\"1000\"/>\n","\n","**Discuss: Explain why toast and frosted mini-wheats got placed near each other in the clustered dataset? Why did ```is_vegan``` and ```is_vegetarian```?**\n","\n","\n","\n","\n","\n","####**Exercise: Run the code below to perform hierarchical clustering on your dataset.**"]},{"cell_type":"code","metadata":{"id":"6nlQSjJFVX8o"},"source":["#@title INSTRUCTOR RESOURCES\n","\n","'''\n","Extra resources for how to explain hierarchical clustering\n","simple explanation: https://www.displayr.com/what-is-hierarchical-clustering/\n","more in depth: https://vitalflux.com/hierarchical-clustering-explained-with-python-example/\n","'''"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"OBULv4cLpw5z"},"source":["f = sns.clustermap(bacteria_counts_lognorm)\n","f.ax_heatmap.set_ylabel('Farm idx')\n","f.ax_heatmap.set_xlabel('Bacteria Species')\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"q78WWh1pTSIn"},"source":["Wow, there's a lot going on here! If you'd like to look at a subset of the data, try using only a small chunk of the data. For example, if you wanted to look at the first 20 farms in the dataset, you would use the line below.\n","\n","\n","```\n","f = sns.clustermap(bacteria_counts_lognorm[:20])\n","```\n","Try a different number in the cell above!\n"]},{"cell_type":"markdown","metadata":{"id":"P5JIVLEipj1-"},"source":["####**Exercise: Answer the following questions:**\n","1. What does it mean if two bacteria got clustered nearby each other?\n","2. Are there any outlier features or samples?\n","3. Can you think of any other fields that might use unsupervised learning? (Hint: Spotify is very famous for its unsupervised learning methods)\n"]},{"cell_type":"code","metadata":{"cellView":"form","id":"-eotAajpgeoR"},"source":["_1_ = '' #@param {type:\"string\"}\n","_2_ = '' #@param {type:\"string\"}\n","_3_ = '' #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"usmd41qsTjQ5"},"source":["# OPTIONAL: Measuring Bacterial Composition\n","\n","How did we get the measurements of our different bacteria in the first place?  This section explores the modern experimental and computational methods that scientists in order to measure bacterial abundance."]},{"cell_type":"markdown","metadata":{"id":"ImHssp5aTqNY"},"source":["## DNA Sequencing\n","\n","Nowadays, measuring the amounts of different species of bacteria in a sample is done through a type of *DNA sequencing* called **16S sequencing**.\n","\n","\n","**Quick Question: What is DNA? If you haven't taken biology recently, ask one of your teammates!  Team work makes the dream work.**\n","\n","DNA sequencing involves using a special machine that can take a piece of DNA and figure out the order of the nucleotides/nitrogenous bases (the building blocks of DNA) that the DNA is composed of.\n","\n","<img src=\"https://www.albert.io/blog/wp-content/uploads/2016/04/Nucleotide.png\" alt=\"drawing\" width=\"300\" >\n","\n","**Quick Question: What are the 4 nucleotides that make up DNA? Again, refer to whoever is the resident biology expert if you can't remember!**\n","\n","\n","If you want to learn more about sequencing on your own time, check out [this article](https://www.khanacademy.org/science/high-school-biology/hs-molecular-genetics/hs-biotechnology/a/dna-sequencing).\n","\n","\n"]},{"cell_type":"markdown","source":["## Instructor Solution\n","<details><summary>click to reveal!</summary\n","\n","Adenine (A)\n","Thymine (T)\n","Cytosine (C)\n","Guanine (G)"],"metadata":{"id":"cytXsL6SdtB3"}},{"cell_type":"markdown","metadata":{"id":"Rv6k_HQvTnK2"},"source":["## The 16S Barcode"]},{"cell_type":"markdown","metadata":{"id":"kXdELCrrV3Bd"},"source":["Bacteria have a very special region in their DNA called the **16S region**. Every species of bacteria has a different DNA sequence in their 16S region, and scientists have created a huge catalog of the 16S sequence in thousands of species of bacteria.\n","\n","\n","The 16S region acts like a \"barcode.\"  Let's give a little metaphor. During 16S sequencing:\n","1.  To measure the amount of bacteria in a sample (analagous to \"the shopping cart\" in this extended metaphor), all of the bacteria are put into a test tube (onto the register at the grocery store) with some special chemicals.\n","2.  The sequencing machine (the cashier's barcode scanner) reads the 16S sequence (the barcode!) of every bacterial cell in the sample and logs it in the computer.\n","3.  Then the computer looks up what species of bacteria (grocery item) each 16S sequence/barcode belongs to, and tallies up the total number of each bacteria.\n","\n","\n","P.S. DNA sequencing is getting drastically more efficient, so sequencing your sample might actually be cheaper and faster than your grocery run these days :)\n","\n","<img src=\"https://static.scientificamerican.com/sciam/cache/file/354C5183-DB53-432A-92B3FC4BA3E99297_source.jpg\n","\" width=500>\n","\n"]},{"cell_type":"markdown","metadata":{"id":"r8LCn_TWSFMq"},"source":["<img src=\"https://upload.wikimedia.org/wikipedia/commons/b/bd/DNA_Barcoding.png\" width=500>\n"]},{"cell_type":"markdown","metadata":{"id":"guvZX7ZES2OE"},"source":["\n","\n","---\n","\n","\n","## From Barcodes to Bacterial Counts\n","\n"]},{"cell_type":"markdown","metadata":{"id":"VDbuEZdri0Ks"},"source":["Let's look at what our ```sequences_counts``` dataframe looks like right now. **Discuss: what are the rows and what are the columns?**"]},{"cell_type":"code","metadata":{"id":"_jofO3VlejMf"},"source":["sequences_counts.head()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"3_dbKifWjCq3"},"source":["#@title ##### We want to get our counts in terms of different bacteria (remember that each sequences maps to a type of bacteria). Something that looks like this (run this cell):\n","bacteria_counts[bacteria_counts.columns[20::20]].head()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MeT72orEjlkG"},"source":["We are going to perform an algorithm in 3 steps.\n","1. Label each sequence (barcode) with the bacteria that corresponds to that particular barcode. You can look this up using the dictionary ```sequence_to_species_dict```.\n","2. Group barcodes/sequences that correspond to the same bacteria species together using the ```pandas``` ```groupby``` function.\n","3. Use the ```pandas``` ```.sum()``` function to sum together the counts of all sequences that correspond to the same bacteria species.\n","\n","Fill in the code below to perform this algorithm."]},{"cell_type":"code","metadata":{"id":"LVWtAah7kNvc"},"source":["# The groupby function works better when we are aggregating over columns,\n","# so we transpose our data.\n","sequences_counts_t = sequences_counts.transpose()\n","\n","# Step 1: Create a new column called \"species\" that corrsponds to the species of a given barcode,\n","# per looked up in the dictionary.\n","sequences_counts_t['### FILL IN ###'] = [sequence_to_species_dict[i] for i in sequences_counts_t.index]\n","\n","\n","# Step 2: Use the pandas \"groupby\" function to group sequences together by species.\n","# and Step 3: Use the .sum() function to sum together the counts of the grouped sequences.\n","summed_data = sequences_counts_t.groupby('### FILL IN ###').sum()\n","\n","# Finally, we will re-transpose the data so that our columns are bacteria and our rows are each farm.\n","bacteria_counts = summed_data.transpose()\n","bacteria_counts.head()"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"C5SsBqjIAJBm"},"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"6Ap3VJFZliij","cellView":"form"},"source":["#@title #### Example Solution\n","# The groupby function works better when we are aggregating over columns,\n","# so we transpose our data.\n","sequences_counts_t = sequences_counts.transpose()\n","\n","# Step 1: Create a new column called \"species\" that corrsponds to the species of a given barcode,\n","# per looked up in the dictionary.\n","sequences_counts_t['species'] = [sequence_to_species_dict[i] for i in sequences_counts_t.index]\n","\n","\n","# Step 2: Use the pandas \"groupby\" function to group sequences together by species.\n","# and Step 3: Use the .sum() function to sum together the counts of the grouped sequences.\n","\n","summed_data = sequences_counts_t.groupby('species').sum()\n","\n","# Finally, we will re-transpose the data so that our columns are bacteria and our rows are each farm.\n","bacteria_counts = summed_data.transpose()\n","bacteria_counts.head()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gpQ-Ya6cmsKH"},"source":["Run the next cell and **discuss: what do you notice about the shape of the ```sequence_counts``` dataframe and the ```bacteria_counts``` dataframe? Why are the shapes different?**"]},{"cell_type":"code","metadata":{"id":"Z5utC-Fzm1zo"},"source":["print('Shape of sequence_counts:', np.shape(sequences_counts))\n","print('Shape of bacteria_counts:', np.shape(bacteria_counts))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"W8QjT0mpnNAN"},"source":["bacteria_counts"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"j0xx1b7cLoDs"},"source":["Now you're able to see how scientists measure the amounts of different bacterial species!\n","\n","In the next notebook, you're actually going to write an algorithm that turns a list of the 16S barcodes output by the sequencer/barcode scanner into counts of different bacteria types. You will use this data to build a machine learning model to predict crop yield from bacterial composition of soil!"]}]}